### PR TITLE
Add Audio Spectogram Transformer

### LINKED ISSUE #16383: AST: Audio Spectrogram Transformer
# 🌟 New model addition

## Model description

In the past decade, convolutional neural networks (CNNs)
have been widely adopted as the main building block for endto-end audio classification models, which aim to learn a direct
mapping from audio spectrograms to corresponding labels. To
better capture long-range global context, a recent trend is to
add a self-attention mechanism on top of the CNN, forming a
CNN-attention hybrid model. However, it is unclear whether
the reliance on a CNN is necessary, and if neural networks
purely based on attention are sufficient to obtain good performance in audio classification. In this paper, we answer the question by introducing the Audio Spectrogram Transformer (AST),
the first convolution-free, purely attention-based model for audio classification. We evaluate AST on various audio classification benchmarks, where it achieves new state-of-the-art results of 0.485 mAP on AudioSet, 95.6% accuracy on ESC-50,
and 98.1% accuracy on Speech Commands V2.
Index Terms: audio classification, self-attention

## Open source status

* [x] the model implementation is available: https://github.com/YuanGongND/ast
* [x] the model weights are available: https://github.com/YuanGongND/ast#Pretrained-Models
* [x] who are the authors: @YuanGongND

Happy to supervise anyone interested in porting the model :-)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Fixes #16383 

This PR adds the [Audio Spectogram Transformer (AST)](https://arxiv.org/abs/2104.01778) model from MIT.

Similar to Whisper (actually prior to Whisper), the model treats audio as an image and applies a Vision Transformer on it.

The model gets SOTA results on audio classification benchmarks.

### CODE REVIEW

File: src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.config = config
+
+    def get_shape(self, config):
+        frequency_stride = config.frequency_stride
+        time_stride = config.time_stride
+        frequency_dimension = config.frequency_dimension
+        time_dimension = config.time_dimension
+        test_input = torch.randn(1, 1, frequency_dimension, time_dimension)
+        test_projection = nn.Conv2d(
```
Comment (sanchit-gandhi) [inline src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py:82]: Is it strictly necessary to perform a 'dummy' convolution to get the downsampled feature dimensions? We can mathematically compute the number of downsampled dims from a conv layer given the kernel size and stride as:
$$W_{out} = \frac{W_{in}−F+2P}{S+1}$$
where $W_{in}$ is the number of input features, $W_{out}$ is the number of output features, $F$ the kernel size, $P$ the amount of zero-padding, and $S$ the stride (https://cs231n.github.io/convolutional-networks/#conv).

It seems a bit wasteful having to do a dummy fprop to compute $W_{out}$ when we can do it analytically!

File: src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+            pooler_output=pooled_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
+
+
+@add_start_docstrings(
+    """
+    Audio Spectogram Transformer model with an audio classification head on top (a linear layer on top of the final
+    hidden state of the [CLS] token) e.g. for AudioSet.
```
Comment (sanchit-gandhi) [inline src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py:541]: This e.g. seems a bit out of place? Is it meant to be "e.g. for sequence classification tasks"?

File: src/transformers/models/audio_spectogram_transformer/test.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import torchaudio
```
Comment (sanchit-gandhi) [inline src/transformers/models/audio_spectogram_transformer/test.py:1]: Is this file part of the PR?

File: tests/models/audio_spectogram_transformer/test_feature_extraction_audio_spectogram_transformer.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+        feature_extractor = self.feature_extraction_class(**self.feat_extract_tester.prepare_feat_extract_dict())
+        np_speech_inputs = np.random.rand(100).astype(np.float64)
+        py_speech_inputs = np_speech_inputs.tolist()
+
+        for inputs in [py_speech_inputs, np_speech_inputs]:
+            np_processed = feature_extractor.pad([{"input_values": inputs}], return_tensors="np")
+            self.assertTrue(np_processed.input_values.dtype == np.float32)
+            pt_processed = feature_extractor.pad([{"input_values": inputs}], return_tensors="pt")
+            self.assertTrue(pt_processed.input_values.dtype == torch.float32)
```
Comment (sanchit-gandhi) [inline tests/models/audio_spectogram_transformer/test_feature_extraction_audio_spectogram_transformer.py:135]: Would be great to have an integration test as well! Similar to https://github.com/huggingface/transformers/blob/f38a1454181d0ac4ef94ba76dfb3c0dcdf45d2b7/tests/models/whisper/test_feature_extraction_whisper.py#L210

File: tests/models/audio_spectogram_transformer/test_modeling_audio_spectogram_transformer.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        self.attention_probs_dropout_prob = attention_probs_dropout_prob
+        self.type_sequence_label_size = type_sequence_label_size
+        self.initializer_range = initializer_range
+        self.scope = scope
+        self.frequency_stride = frequency_stride
+        self.time_stride = time_stride
+
+        # in AudioSpectogramTransformer, the seq length equals the number of patches + 2 (we add 2 for the [CLS] and distillation tokens)
+        test_input = torch.randn(1, 1, self.frequency_dimension, self.time_dimension)
+        test_projection = nn.Conv2d(
```
Comment (sanchit-gandhi) [inline tests/models/audio_spectogram_transformer/test_modeling_audio_spectogram_transformer.py:90]: Again, don't think it's strictly necessary to have to perform this dummy conv operation!

File: src/transformers/models/audio_spectogram_transformer/feature_extraction_audio_spectogram_transformer.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        **kwargs
+    ) -> BatchFeature:
+        """
+        Main method to featurize and prepare for the model one or several sequence(s).
+
+        Args:
+            raw_speech (`np.ndarray`, `List[float]`, `List[np.ndarray]`, `List[List[float]]`):
+                The sequence or batch of sequences to be padded. Each sequence can be a numpy array, a list of float
+                values, a list of numpy arrays or a list of list of float values.
```
Comment (sanchit-gandhi) [inline src/transformers/models/audio_spectogram_transformer/feature_extraction_audio_spectogram_transformer.py:136]: Wonder whether it makes sense to adapt this to handle torch tensors as inputs? I know we don't do this currently in Speech2Text2, but I saw that in quite a few of the tests we go from: torch -> numpy -> feat-extractor -> torch, which seems pretty convoluted!

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for adding this audio-classification model @NielsRogge! The modelling components in general look good to me 👌 A couple of questions regarding the feature extractor (namely the mean / std normalisation). Also, it would be great to document the performant audio-classification capabilities of this model with an inference/fine-tuning example at [examples/pytorch/audio-classification](https://github.com/huggingface/transformers/tree/main/examples/pytorch/audio-classification)! 🚀 Very excited to see the accompanying tutorial video for this model addition 😉

Also one thing you're gonna kick yourself for... it's "spect**r**ogram" not "spectogram" 😅 hope the renaming isn't too arduous!

File: src/transformers/models/audio_spectogram_transformer/feature_extraction_audio_spectogram_transformer.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+            The sampling rate at which the audio files should be digitalized expressed in Hertz per second (Hz).
+        num_mel_bins (`int`, defaults to 128):
+            Number of Mel-frequency bins.
+        do_normalize (`bool`, *optional*, defaults to `True`):
+            Whether or not to apply utterance-level cepstral mean and variance normalization to extracted features.
+        mean (`int`, *optional*, defaults to -4.2677393):
+            Whether or not to normalize the extracted features to a mean of 0. Uses the AudioSet mean by default,
+            obtained by the authors.
+        std (`int`, *optional*, defaults to 4.5689974):
+            Whether or not to normalize the extracted features to a std of 05. Uses the AudioSet std by default,
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/audio_spectogram_transformer/feature_extraction_audio_spectogram_transformer.py]: Yes that's correct, will update

File: docs/source/en/model_doc/audio-spectogram-transformer.mdx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+Tips:
+
+- When fine-tuning the Audio Spectogram Transformer (AST) on your own dataset, it's recommended to take care of the input normalization.
+The authors normalize the input audio to have a 0 mean and 0.5 std. To use the pretrained model, you should roughly normalize the input to
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply docs/source/en/model_doc/audio-spectogram-transformer.mdx]: It's indeed just computing a single mean and std for your downstream dataset, and then normalize all training examples you have using this mean and std. It's similar to image classification, where you normalize all images using the ImageNet mean + std.

File: src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py (hunk 2 above)
Comment (NielsRogge) [reply src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py]: Could you clarify here? It's meant for audio classification tasks. But I saw all audio classifiers are called `xxxForSequenceClassification` in the library atm.

File: src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py]: Thanks, indeed! Used this formula now.

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+AUDIO_SPECTROGRAM_TRANSFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    # TODO update to appropriate organization
+    "nielsr/audio-spectrogram-transformer-finetuned-audioset-10-10-0.4593": "https://huggingface.co/nielsr/audio-spectrogram-transformer-finetuned-audioset-10-10-0.4593/resolve/main/config.json",
```
Comment (sgugger) [inline src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:26]: The checkpoint name is starting to be a bit long, would it make sense to abbreviate audio-spectogram-transformer to ast?

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+    Example:
+
+    ```python
+    >>> from transformers import AudioSpectrogramTransformerModel, AudioSpectrogramTransformerConfig
+
+    >>> # Initializing a AudioSpectrogramTransformer audio_spectrogram_transformer-base-patch16-224 style configuration
+    >>> configuration = AudioSpectrogramTransformerConfig()
+
+    >>> # Initializing a model from the audio_spectrogram_transformer-base-patch16-224 style configuration
```
Comment (sgugger) [inline src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:83]: You should use the new version for the config docstring here (look at BERT for an example).

File: src/transformers/models/audio_spectrogram_transformer/feature_extraction_audio_spectrogram_transformer.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    This class extracts mel-filter bank features from raw speech using TorchAudio and applies utterance-level cepstral
+    mean and variance normalization to the extracted features.
+
+    Args:
+        feature_size (`int`, defaults to 1):
+            The feature dimension of the extracted features.
+        sampling_rate (`int`, defaults to 16000):
+            The sampling rate at which the audio files should be digitalized expressed in Hertz per second (Hz).
+        num_mel_bins (`int`, defaults to 128):
```
Comment (sgugger) [inline src/transformers/models/audio_spectrogram_transformer/feature_extraction_audio_spectrogram_transformer.py:48]: Those are all optionals since they have a default :-)

**Review by sgugger — APPROVED** [review]
Very clean, thanks a lot for adding this new model!

**Comment by FrankFundel** [comment]
Hey, how can I use it already? I installed the branch, but unsure how to load the model. I'm new with huggingface :D

File: docs/source/en/model_doc/audio-spectogram-transformer.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply docs/source/en/model_doc/audio-spectogram-transformer.mdx]: Okay great! So to summarise:
- Training: compute the mean/std over your fine-tuning dataset
- Inference: use the expected mean/std from the AudioSet dataset (with the default values in the feature extractor)

=> perhaps we can clarify this in the .md with a similar summary?

File: src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py]: Sorry! Meant the `e.g. ...` bit. I can't GH suggest as the comment is outdated, but I'd simply write:

> Audio Spectogram Transformer model with an audio classification head on top (a linear layer on top of the final
    hidden state of the [CLS] token)

**Comment by sanchit-gandhi** [comment]
Hey @FrankFundel - hoping @NielsRogge adds a nice example as part of this PR documenting just that 🤞 In the mean time, you can try adapting the example from https://huggingface.co/docs/transformers/tasks/audio_classification

You'll need to change the repo names from `facebook/wav2vec2-base` to the appropriate Audio Spectrogram Transformer repo name. You'll also need to change the preprocess function (https://huggingface.co/docs/transformers/tasks/audio_classification#preprocess) to something like:

```python
def preprocess_function(examples):
    audio_arrays = [x["array"] for x in examples["audio"]]
    input_features = feature_extractor(audio_array, sampling_rate=feature_extractor.sampling_rate)
    return input_features
```
This is all currently untested, so might require some playing around to make it work.

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py (hunk 9 above)
Comment (NielsRogge) [reply src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:26]: Ok I'll update to `ast`!
